Move watcher tracking strategies to separate classes

wolfs opened the pull request:
and add some tests for them.

lptr commented on subprojects/core/src/main/java/org/gradle/internal/vfs/AbstractEventDrivenFileWatcherRegistry.java:
```diff
@@ -16,53 +16,86 @@
-    public AbstractEventDrivenFileWatcherRegistry(FileWatcherCreator watcherCreator, ChangeHandler handler) {
+    public AbstractEventDrivenFileWatcherRegistry(FileWatcherCreator watcherCreator, ChangeHandler handler, Function<FileWatcher, FileWatcherUpdater> watcherUpdaterCreator) {
         this.watcher = createWatcher(watcherCreator);
+        this.fileWatcherUpdater = watcherUpdaterCreator.apply(watcher);
```
We used to have the registry be responsible for the creation of the callback, but that's not necessary anymore. We could pass in the event queue and the watcher here. That way we could keep the updater separate from this class completely.

wolfs replied:
I removed the `WatcherCreator` interface. The corresponding factories now take care of instantiating the right things.

lptr commented on subprojects/core/src/main/java/org/gradle/internal/vfs/LinuxFileWatcherUpdater.java:
```diff
@@ -0,0 +1,172 @@
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+public class LinuxFileWatcherUpdater implements FileWatcherUpdater {
```
Should this be called `Linux` or `NonHierarchical`?

wolfs replied:
I renamed it to `NonHierarchical`.

lptr commented on subprojects/core/src/test/groovy/org/gradle/internal/vfs/AbstractFileWatcherUpdaterTest.groovy:
```diff
@@ -0,0 +1,136 @@
+    abstract FileWatcherUpdater createUpdater(FileWatcher watcher)
+
+    def "starts and stops watching must watch directories"() {
+        def mustWatchDirectories = ["first", "second", "third"].collect { temporaryFolder.testDirectory.createDir(it)}
```
Nit:
```suggestion
        def mustWatchDirectories = ["first", "second", "third"].collect { temporaryFolder.testDirectory.createDir(it) }
```

wolfs replied:
Fixed.

lptr commented on subprojects/core/src/test/groovy/org/gradle/internal/vfs/AbstractFileWatcherUpdaterTest.groovy:
```diff
@@ -0,0 +1,133 @@
+
+    static boolean equalAsSets(Object actual, Object expected) {
+        return (actual as Set) == (expected as Set)
+    }
```
This is problematic. I think we should be able to declare the type of the parameters more strictly.

I would also prefer if instead of converting to `Set` we'd compare a sorted list of elements. That way we'd be able to spot if an element appears multiple times (which it shouldn't).

wolfs replied:
I used a collection first, though then IDEA was not able to infer the argument types from where this method has been used. That is why I ended up with `Object` in the end...

wolfs replied:
Now the second argument is a `Collection` and the method compares the sorted collections.

lptr commented on subprojects/core/src/test/groovy/org/gradle/internal/vfs/HierarchicalFileWatcherUpdaterTest.groovy:
```diff
@@ -0,0 +1,86 @@
+        then:
+        1 * watcher.startWatching({ equalAsSets(it, mustWatchDirectoryRoots) })
+        0 * _
+    }
```
Do we need a test for "upgrading" a watch, i.e. when `first/within` is watched first, and then we need to start `first`, we'd need to unregister `first/within`, right?

wolfs replied:
I added a test for the use-case you are describing.

lptr replied:
I think this is fine, though it would be more readable for me if these new cases were added as separate tests instead.

wolfs replied:
I did split the test now.

lptr commented on subprojects/core/src/test/groovy/org/gradle/internal/vfs/LinuxFileWatcherUpdaterTest.groovy:
```diff
@@ -0,0 +1,64 @@
+
+    def "adds watches for all must watch directories"() {
+        def mustWatchDirectoryRoots = ["first", "second"].collect { file(it).createDir()}
+        def mustWatchDirectories = mustWatchDirectoryRoots + file("first/within").createDir()
```
Nit: splitting this into two separate variables does not seem to add anything for me, can we merge them?

wolfs replied:
I am using the separate variables, so I can easily assert which parts we should watch and which we shouldn't.

lptr replied:
Alright.

wolfs commented:
@lptr PTAL!

lptr commented on subprojects/core/src/main/java/org/gradle/internal/vfs/AbstractFileWatcherRegistryFactory.java:
```diff
@@ -0,0 +1,50 @@
+
+public abstract class AbstractFileWatcherRegistryFactory implements FileWatcherRegistryFactory {
+    @Override
+    public FileWatcherRegistry startWatcher(FileWatcherRegistry.ChangeHandler handler) {
```
Should this method be called `createFileWatcherRegistry()` or `startWatcherRegistry()` or somesuch?

wolfs replied:
Renamed to `createFileWatcherRegistry`.

lptr commented on subprojects/core/src/main/java/org/gradle/internal/vfs/AbstractFileWatcherRegistryFactory.java:
```diff
@@ -0,0 +1,50 @@
+public abstract class AbstractFileWatcherRegistryFactory implements FileWatcherRegistryFactory {
+    @Override
+    public FileWatcherRegistry startWatcher(FileWatcherRegistry.ChangeHandler handler) {
+        BlockingQueue<FileWatchEvent> fileEvents = new ArrayBlockingQueue<>(4096);
```
Please extract magic number to named constant.

wolfs replied:
Extracted.

lptr commented on subprojects/core/src/main/java/org/gradle/internal/vfs/EventDrivenFileWatcherRegistry.java:
```diff
@@ -28,31 +27,30 @@
 
-public abstract class AbstractEventDrivenFileWatcherRegistry implements FileWatcherRegistry {
-    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractEventDrivenFileWatcherRegistry.class);
+public class EventDrivenFileWatcherRegistry implements FileWatcherRegistry {
```
Why not `DefaultFileWatcherRegistry`? Do we expect more implementations of this?

wolfs replied:
`DefaultFileWatcherRegistry` it is.

lptr commented on subprojects/core/src/main/java/org/gradle/internal/vfs/EventDrivenFileWatcherRegistry.java:
```diff
@@ -28,31 +27,30 @@
+    public EventDrivenFileWatcherRegistry(FileWatcher watcher, ChangeHandler handler, FileWatcherUpdater fileWatcherUpdater, BlockingQueue<FileWatchEvent> fileEvents) {
+        this.watcher = watcher;
+        this.fileEvents = fileEvents;
+        this.fileWatcherUpdater = fileWatcherUpdater;
```
This class now just keeps hold of this instance to return later. That doesn't seem good. I wonder if we could keep track of it in some better way. Or maybe it would be better if `FileWatcherRegistry` delegated to the updater. Maybe even wrap `FileWatcherRegistry` in an `MutableFileWatcherRegistry` to add the necessary methods.

wolfs replied:
Currently, the watcher registry holds together various parts, which include the updater. If we want to move the updater out of here, then we need something else to hold it or to query the watcher from the registry. Both things don't seem very appealing to me.

Let's revisit this when we re-organized the project structure.

lptr commented:
I don't know where that comment went, but I'd still like to see `FileWatcherUpdater.changed()` have a better name. I think it shouldn't implement `SnapshotDiffListener`.

wolfs commented:
@lptr PTAL!

> I don't know where that comment went, but I'd still like to see FileWatcherUpdater.changed() have a better name. I think it shouldn't implement SnapshotDiffListener.

Let's talk about this when we moved the things to a new subproject.
